File: src/main/java/rr/language/RRFoldingBuilder.java
package rr.language;

import com.intellij.lang.ASTNode;
import com.intellij.lang.folding.FoldingBuilderEx;
import com.intellij.lang.folding.FoldingDescriptor;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.project.DumbAware;
import com.intellij.psi.PsiElement;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rr.language.psi.*;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class RRFoldingBuilder extends FoldingBuilderEx implements DumbAware {

    private static void fold(@NotNull PsiElement e,
                             @Nullable PsiElement l,
                             @Nullable PsiElement r,
                             @NotNull String placeholderText,
                             @NotNull List<FoldingDescriptor> result) {
        if (l != null && r != null) {
            result.add(new FoldingDescriptor(e, l.getTextRange().getStartOffset(), r.getTextRange().getEndOffset(), null, placeholderText));
        }
    }

    @Override
    public @NotNull FoldingDescriptor[] buildFoldRegions(@NotNull PsiElement root, @NotNull Document document, boolean quick) {

        RRFile file = (RRFile) root;
        final List<FoldingDescriptor> folds = new ArrayList<>();

        if (!quick) {
            // Comment folding
            Set<PsiElement> processedComments = new HashSet<>();
            PsiTreeUtil.processElements(file, element -> {
                // Comment folding
                if (element.getNode().getElementType() == RRTypes.COMMENT) {
                    PsiElement end = findFoldingEndForComment(element, processedComments);

                    if (end != null) {
                        folds.add(new FoldingDescriptor(
                            element,
                            element.getTextRange().getStartOffset(),
                            end.getTextRange().getEndOffset(),
                            null,
                            (element.getText().length() > 20 ? element.getText().substring(0, 20) : element.getText()).concat("...")
                        ));
                    }
                }
                return true;
            });

            // KV format folding
            for (RRKVFValue element : PsiTreeUtil.findChildrenOfType(file, RRKVFValue.class)) {
                fold(element, element.getFirstChild(), element.getLastChild(), "...", folds);
            }

            // descr_strat file
            // Sections
            for (RRCampaignSection element : PsiTreeUtil.findChildrenOfType(file, RRCampaignSection.class)) {
                fold(
                    element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            for (RRLandmarkSection element : PsiTreeUtil.findChildrenOfType(file, RRLandmarkSection.class)) {
                // Section might have no children.
                if (element.getFirstChild() != null) {
                    fold(
                        element, element.getFirstChild().getNextSibling(),
                        element.getLastChild(),
                        "...",
                        folds
                    );
                }
            }

            for (RRResourceQuantityEnabled_ element : PsiTreeUtil.findChildrenOfType(file, RRResourceQuantityEnabled_.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            for (RRResourceQuantityDisabled_ element : PsiTreeUtil.findChildrenOfType(file, RRResourceQuantityDisabled_.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            for (RRFactionItem element : PsiTreeUtil.findChildrenOfType(file, RRFactionItem.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            for (RRSettlementItem element : PsiTreeUtil.findChildrenOfType(file, RRSettlementItem.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), " ".concat(element.name()), folds);
            }

            for (RRCharacterItem element : PsiTreeUtil.findChildrenOfType(file, RRCharacterItem.class)) {
                fold(element, element.getCharacterType().getNextSibling(), element.getLastChild(), "...", folds);
            }

            for (RRDiplomacySection element : PsiTreeUtil.findChildrenOfType(file, RRDiplomacySection.class)) {
                fold(element, element.getFirstChild(), element.getLastChild(), "DIPLOMACY SECTION...", folds);
            }

            for (RRSuperfactionSetupItem element : PsiTreeUtil.findChildrenOfType(file, RRSuperfactionSetupItem.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // export_descr_unit file
            for (RRUnitItem_ element : PsiTreeUtil.findChildrenOfType(file, RRUnitItem_.class)) {
                fold(element, element.getNode().findChildByType(RRTypes.DICTIONARY).getPsi(), element.getLastChild(), "...", folds);
            }

            // export_descr_buildings
            for (RRBuildingTree element : PsiTreeUtil.findChildrenOfType(file, RRBuildingTree.class)) {
                fold(element, element.getNode().findChildByType(RRTypes.OCB).getPsi(), element.getLastChild(), "...", folds);
            }

            for (RRBuildingLevel element : PsiTreeUtil.findChildrenOfType(file, RRBuildingLevel.class)) {
                fold(element, element.getNode().findChildByType(RRTypes.REQUIREMENT).getPsi(), element.getLastChild(), "...", folds);
            }

            // export_descr_ancillaries
            for (RRAncillaryDef element : PsiTreeUtil.findChildrenOfType(file, RRAncillaryDef.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            for (RRAncillaryTriggerDef element : PsiTreeUtil.findChildrenOfType(file, RRAncillaryTriggerDef.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // export_descr_traits
            for (RRTraitDef element : PsiTreeUtil.findChildrenOfType(file, RRTraitDef.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            for (RRTraitTriggerDef element : PsiTreeUtil.findChildrenOfType(file, RRTraitTriggerDef.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // descr_mercenaries
            for (RRMercenaryPool element : PsiTreeUtil.findChildrenOfType(file, RRMercenaryPool.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // descr_regions
            for (RRRegionDef element : PsiTreeUtil.findChildrenOfType(file, RRRegionDef.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // descr_names
            for (RRNamesForFaction element : PsiTreeUtil.findChildrenOfType(file, RRNamesForFaction.class)) {
                fold(
                    element,
                    element.getFirstChild().getNextSibling().getNextSibling().getNextSibling().getNextSibling(),
                    element.getLastChild(),
                    "...",
                    folds
                );
            }

            // descr_unit_variation
            for (RRSkin_ element : PsiTreeUtil.findChildrenOfType(file, RRSkin_.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RREthnicity_ element : PsiTreeUtil.findChildrenOfType(file, RREthnicity_.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RREthnicityMakeup_ element : PsiTreeUtil.findChildrenOfType(file, RREthnicityMakeup_.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // feral_descr_ai_personality file
            for (RRAiBuildingPriority element : PsiTreeUtil.findChildrenOfType(file, RRAiBuildingPriority.class)) {
                fold(element, element.getId().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRAiMilitaryPriority element : PsiTreeUtil.findChildrenOfType(file, RRAiMilitaryPriority.class)) {
                fold(element, element.getId().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRAiDiplomaticPriority element : PsiTreeUtil.findChildrenOfType(file, RRAiDiplomaticPriority.class)) {
                fold(element, element.getId().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRAiPersonality element : PsiTreeUtil.findChildrenOfType(file, RRAiPersonality.class)) {
                fold(element, element.getAiPersonalityNameDecl().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_faction_groups file
            for (RRFactionGroup element : PsiTreeUtil.findChildrenOfType(file, RRFactionGroup.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_sm_factions file
            for (RRFactionDecl element : PsiTreeUtil.findChildrenOfType(file, RRFactionDecl.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_sm_resources file
            for (RRResourceDecl element : PsiTreeUtil.findChildrenOfType(file, RRResourceDecl.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // feral_descr_portraits_variation file
            for (RRCulturePortraitVariation element : PsiTreeUtil.findChildrenOfType(file, RRCulturePortraitVariation.class)) {
                fold(element, element.getFirstChild().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_banners file
            for (RRCultureBanner element : PsiTreeUtil.findChildrenOfType(file, RRCultureBanner.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRFactionBanner element : PsiTreeUtil.findChildrenOfType(file, RRFactionBanner.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_character file
            for (RRCharacterTypeDef element : PsiTreeUtil.findChildrenOfType(file, RRCharacterTypeDef.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRCharacterTypeFactionDef element : PsiTreeUtil.findChildrenOfType(file, RRCharacterTypeFactionDef.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }

            // descr_building_battle file
            for (RRTextureVariants_ element : PsiTreeUtil.findChildrenOfType(file, RRTextureVariants_.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRTextureVariant element : PsiTreeUtil.findChildrenOfType(file, RRTextureVariant.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRSpotItems_ element : PsiTreeUtil.findChildrenOfType(file, RRSpotItems_.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRSpotItem element : PsiTreeUtil.findChildrenOfType(file, RRSpotItem.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRStatCategories element : PsiTreeUtil.findChildrenOfType(file, RRStatCategories.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRStatCategory element : PsiTreeUtil.findChildrenOfType(file, RRStatCategory.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRTransitionScripts_ element : PsiTreeUtil.findChildrenOfType(file, RRTransitionScripts_.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRTransitionScript element : PsiTreeUtil.findChildrenOfType(file, RRTransitionScript.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
            for (RRConstructionSite element : PsiTreeUtil.findChildrenOfType(file, RRConstructionSite.class)) {
                fold(element, element.getFirstChild().getNextSibling().getNextSibling(), element.getLastChild(), "...", folds);
            }
        }

        return folds.toArray(new FoldingDescriptor[folds.size()]);
    }

    @Override
    public @Nullable String getPlaceholderText(@NotNull ASTNode node) {
        return null;
    }

    @Override
    public boolean isCollapsedByDefault(@NotNull ASTNode node) {
        return false;
    }

    private static PsiElement findFoldingEndForComment(@NotNull PsiElement comment, @NotNull Set<PsiElement> processedComments) {
        if (processedComments.contains(comment)) return null;

        PsiElement next = comment.getNextSibling();
        PsiElement end = null;

        while (next != null) {
            IElementType elementType = next.getNode().getElementType();

            if (elementType != RRTypes.COMMENT && elementType != TokenType.WHITE_SPACE) {
                break;
            }

            if (elementType == RRTypes.COMMENT) {
                end = next;
                processedComments.add(next);
            }

            next = next.getNextSibling();
        }

        return end;
    }
}

File: src/main/java/rr/language/RRFoldingBuilder.kt
package rr.language

import com.intellij.lang.ASTNode
import com.intellij.lang.folding.FoldingBuilderEx
import com.intellij.lang.folding.FoldingDescriptor
import com.intellij.openapi.editor.Document
import com.intellij.openapi.project.DumbAware
import com.intellij.psi.PsiElement
import com.intellij.psi.TokenType
import com.intellij.psi.util.PsiTreeUtil
import rr.language.psi.*
import rr.language.psi.impl.RROffmapWonderImpl

class RRFoldingBuilder : FoldingBuilderEx(), DumbAware {
    override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<FoldingDescriptor> {
        val file = root as RRFile
        val folds: MutableList<FoldingDescriptor> = ArrayList()
        if (quick) {
            return folds.toTypedArray<FoldingDescriptor>()
        }

        // Comment folding
        val processedComments: MutableSet<PsiElement> = HashSet()
        PsiTreeUtil.processElements(file) { element: PsiElement ->
            // Comment folding
            if (element.node.elementType === RRTypes.COMMENT) {
                val end = findFoldingEndForComment(element, processedComments)
                if (end != null) {
                    folds.add(
                        FoldingDescriptor(
                            element,
                            element.textRange.startOffset,
                            end.textRange.endOffset,
                            null,
                            (if (element.text.length > 20) element.text.substring(0, 20) else element.text) + "..."
                        )
                    )
                }
            }
            true
        }

        // KV format folding
        for (element in PsiTreeUtil.findChildrenOfType(file, RRKVFValue::class.java)) {
            fold(element, element.firstChild, element.lastChild, "...", folds)
        }

        // descr_strat file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRCampaignSection::class.java)) {
            fold(
                element, element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRLandmarkSection::class.java)) {
            // Section might have no children.
            if (element.firstChild != null) {
                fold(
                    element, element.firstChild.nextSibling,
                    element.lastChild,
                    "...",
                    folds
                )
            }
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRResourceQuantityEnabled_::class.java
        )) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRResourceQuantityDisabled_::class.java
        )) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRResourcesSection::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRFactionItem::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRSettlementItem::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, " " + element.name(), folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRCharacterItem::class.java)) {
            fold(element, element.characterType.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRDiplomacySection::class.java)) {
            fold(element, element.firstChild, element.lastChild, "DIPLOMACY SECTION...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRSuperfactionSetupItem::class.java
        )) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }

        // export_descr_unit file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRUnitItem_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // export_descr_buildings
        for (element in PsiTreeUtil.findChildrenOfType(file, RRBuildingTree::class.java)) {
            fold(element, element.node.findChildByType(RRTypes.OCB)!!.psi, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRBuildingLevel::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // export_descr_ancillaries
        for (element in PsiTreeUtil.findChildrenOfType(file, RRAncillaryDef::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRAncillaryTriggerDef::class.java
        )) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }

        // export_descr_traits
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTraitDef::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTraitTriggerDef::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }

        // descr_mercenaries
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMercenaryPool::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }

        // descr_regions
        for (element in PsiTreeUtil.findChildrenOfType(file, RRRegionDef::class.java)) {
            fold(
                element,
                element.firstChild.nextSibling.nextSibling,
                element.lastChild,
                "...",
                folds
            )
        }

        // descr_unit_variation
        for (element in PsiTreeUtil.findChildrenOfType(file, RRSkin_::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RREthnicity_::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RREthnicityMakeup_::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // feral_descr_ai_personality file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRAiBuildingPriority::class.java)) {
            fold(element, element.id.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRAiMilitaryPriority::class.java)) {
            fold(element, element.id.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRAiDiplomaticPriority::class.java
        )) {
            fold(element, element.id.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRAiPersonality::class.java)) {
            fold(element, element.aiPersonalityNameDecl.nextSibling, element.lastChild, "...", folds)
        }

        // descr_faction_groups file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRFactionGroup::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_sm_factions file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRFactionDecl::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // descr_sm_resources file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRResourceDecl::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // feral_descr_portraits_variation file
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRCulturePortraitVariation::class.java
        )) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // descr_banners file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRCultureBanner::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRFactionBanner::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_character file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRCharacterTypeDef::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(
            file,
            RRCharacterTypeFactionDef::class.java
        )) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_building_battle file
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTextureVariants_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTextureVariant::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRSpotItems_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRSpotItem::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRStatCategories::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRStatCategory::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTransitionScripts_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRTransitionScript::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRConstructionSite::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_offmap_models
        for (element in PsiTreeUtil.findChildrenOfType(file, RROffmapNavy::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RROffmapSettlement::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RROffmapWonderImpl::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RROffmapPort::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // descr_model_battle, descr_model_strat
        for (element in PsiTreeUtil.findChildrenOfType(file, RRModel_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_mount
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMount_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling.nextSibling, element.lastChild, "...", folds)
        }

        // descr_beliefs
        for (element in PsiTreeUtil.findChildrenOfType(file, RRReligion_::class.java)) {
            fold(element, element.firstChild.nextSibling, element.lastChild, "...", folds)
        }

        // descr_projectiles_new
        for (element in PsiTreeUtil.findChildrenOfType(file, RRProjectile_::class.java)) {
            fold(element, element.projectileNameDecl.nextSibling, element.lastChild, "...", folds)
        }

        // rebel_faction_descr
        for (element in PsiTreeUtil.findChildrenOfType(file, RRRebelFaction::class.java)) {
            fold(element, element.rebelFactionNameDef.nextSibling, element.lastChild, "...", folds)
        }

        // scripts
        for (element in PsiTreeUtil.findChildrenOfType(file, RRIf_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRIfNot_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRWhile_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRWhileNot_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRForEach_::class.java)) {
            fold(element, element.scope!!.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMonitor_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMonitorConditions_::class.java)) {
            fold(element, element.conditionAtomList[0].nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMonitorEvent_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRSpawnBattle_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRBenchmark_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        for (element in PsiTreeUtil.findChildrenOfType(file, RRMacroDef_::class.java)) {
            fold(element, element.firstChild.nextSibling.nextSibling, element.lastChild, "...", folds)
        }
        return folds.toTypedArray<FoldingDescriptor>()
    }

    override fun getPlaceholderText(node: ASTNode): String? {
        return null
    }

    override fun isCollapsedByDefault(node: ASTNode): Boolean {
        return false
    }

    companion object {
        private fun fold(
            e: PsiElement,
            l: PsiElement?,
            r: PsiElement?,
            placeholderText: String,
            result: MutableList<FoldingDescriptor>
        ) {
            if (l != null && r != null) {
                result.add(FoldingDescriptor(e, l.textRange.startOffset, r.textRange.endOffset, null, placeholderText))
            }
        }

        private fun findFoldingEndForComment(
            comment: PsiElement,
            processedComments: MutableSet<PsiElement>
        ): PsiElement? {
            if (processedComments.contains(comment)) return null
            var next = comment.nextSibling
            var end: PsiElement? = null
            while (next != null) {
                val elementType = next.node.elementType
                if (elementType !== RRTypes.COMMENT && elementType !== TokenType.WHITE_SPACE) {
                    break
                }
                if (elementType === RRTypes.COMMENT) {
                    end = next
                    processedComments.add(next)
                }
                next = next.nextSibling
            }
            return end
        }
    }
}
